rfd/0083-prevent-user-role-self-lock.md

```diff
@@ -0,0 +1,133 @@
+---
+authors: Jan Kaczmarkiewicz (jan.kaczmarkiewicz@goteleport.com)
+state: draft
+---
+
+# RFD 83 - Prevent user role self-lock
+
+## Required Approvers
+
+- Engineering: @jimbishopp
+- Product: @xinding33
+
+## What
+
+Users can change the permission in a way that can cause no one to be able to change it further.
+
+Currently, there are multiple ways for users to lock themselves:
+
+| initial state                              | operation                                                                                   | result |
+| ------------------------------------------ | ------------------------------------------------------------------------------------------- | :----: |
+| 1 _admin_ and _regular users_              | _admin_ removes himself from an `editor` role                                               |   🔒   |
+| 1 _admin_ and _regular users_              | _admin_ loses access to own account                                                         |   🔒   |
+| 1 _admin_, _SSO admin_ and _regular users_ | _SSO admin_ removes the `editor` role from _admin_ and SSO connector. _SSO admin_ logs out. |   🔒   |
+| _SSO admin_ and _regular users_            | Someone on the SSO side removes _SSO admin_ user or team/group containing _SSO admins_      |   🔒   |
+
+Cases handled currently:
+
+| initial state               | operation                                                                          | result |
+| --------------------------- | ---------------------------------------------------------------------------------- | :----: |
+| _admin_ and _regular users_ | _admin_ tries to delete his account. Operation is not permitted, an error is shown |   🔓   |
+
+## Terminology
+
+- _admin_: local user with `editor` role
+- _SSO admin_: user with `editor` role added via SSO connector (eg. GitHub)
+- _regular user_: user without `editor` role
+- 🔒: No one can edit roles. Deadlock
+- 🔓: There is still an active user with an editor role
+
+## Why
+
+When a user locks themselves are unable to further manage the teleport cluster. We want to prevent that since this is bad for the user experience and prevents them from using the product.
```

Comment:
Further, for our Cloud offering we don't perform account resets as a policy. The only recurse for a customer is to delete their cluster & deploy a new one. This inevitably leads to data-loss.

On-prem customers could work around this "locked out" state by manually instrumenting the `auth` process using `tctl`.

Reply:
fec4c734ce2d1180ca2f0b5715533d90cc5b0fcd